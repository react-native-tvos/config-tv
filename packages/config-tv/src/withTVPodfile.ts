import { mergeContents } from '@expo/config-plugins/build/utils/generateCode';
import { ConfigPlugin, withDangerousMod } from 'expo/config-plugins';
import { promises } from 'fs';
import path from 'path';

import { ConfigData } from './types';
import { verboseLog } from './utils';

/** Dangerously makes or reverts TV changes in the project Podfile. */
export const withTVPodfile: ConfigPlugin<ConfigData> = (c, params = {}) => {
  return withDangerousMod(c, [
    'ios',
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    async (config) => {
      const file = path.join(config.modRequest.platformProjectRoot, 'Podfile');

      const contents = await promises.readFile(file, 'utf8');

      const modifiedContents = addTVPodfileModifications(contents);

      verboseLog('modifying Podfile for tvOS', {
        params,
        platform: 'ios',
        property: 'podfile',
      });
      await promises.writeFile(file, modifiedContents, 'utf-8');

      return config;
    },
  ]);
};

const MOD_TAG = 'react-native-tvos-import';

export function addTVPodfileModifications(src: string): string {
  if (src.indexOf('platform :tvos') !== -1) {
    return src;
  }
  const newSrc = mergeContents({
    tag: MOD_TAG,
    src,
    newSrc:
      "source 'https://github.com/react-native-tvos/react-native-tvos-podspecs.git'\nsource 'https://cdn.cocoapods.org/'\n",
    anchor: /^/,
    offset: 0,
    comment: '#',
  }).contents;

  return newSrc.replace('platform :ios', 'platform :tvos');
}
