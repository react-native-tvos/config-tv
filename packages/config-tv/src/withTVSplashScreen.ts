import { ConfigPlugin, withDangerousMod } from 'expo/config-plugins';
import { promises } from 'fs';
import path from 'path';

import { ConfigData } from './types';
import { packageNameAndVersion, verboseLog } from './utils';

/** Dangerously makes changes needed for TV in SplashScreen.storyboard. */
export const withTVSplashScreen: ConfigPlugin<ConfigData> = (
  config,
  params = {},
) => {
  return withDangerousMod(config, [
    'ios',
    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    async (config) => {
      if (!config.modRequest.projectName) {
        throw new Error(
          `The ${packageNameAndVersion} plugin requires a configured project name.`,
        );
      }
      const file = path.join(
        config.modRequest.platformProjectRoot,
        config.modRequest.projectName,
        'SplashScreen.storyboard',
      );

      const contents = await promises.readFile(file, 'utf8');

      const modifiedContents = addTVSplashScreenModifications(contents);

      verboseLog('modifying SplashScreen.storyboard for tvOS', {
        params,
        platform: 'ios',
        property: 'splashscreen',
      });
      await promises.writeFile(file, modifiedContents, 'utf-8');

      return config;
    },
  ]);
};

const splashScreenStringsForPhone = [
  'com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB',
  'targetRuntime="iOS.CocoaTouch"',
  'id="retina5_5"',
  '<deployment identifier="iOS"/>',
];

const splashScreenStringsForTV = [
  'com.apple.InterfaceBuilder.AppleTV.Storyboard',
  'targetRuntime="AppleTV"',
  'id="appleTV"',
  '<deployment identifier="tvOS"/>',
];

function modifySource(
  src: string,
  originalStrings: string[],
  replacementStrings: string[],
): string {
  let modifiedSource = src;
  originalStrings.forEach((s, i) => {
    const original = new RegExp(`${s}`, 'g');
    const replacement = replacementStrings[i];
    modifiedSource = modifiedSource.replace(original, replacement);
  });
  return modifiedSource;
}

export function addTVSplashScreenModifications(src: string): string {
  return modifySource(
    src,
    splashScreenStringsForPhone,
    splashScreenStringsForTV,
  );
}
