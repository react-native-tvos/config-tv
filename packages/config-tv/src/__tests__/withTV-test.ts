import { AndroidConfig } from '@expo/config-plugins';
import { promises as fs } from 'fs';
import { vol } from 'memfs';
import { join, resolve } from 'path';

import {
  originalPodfile,
  originalSplashScreen,
  originalAndroidManifest,
  originalAndroidManifestNoMainIntent,
} from './testConstants';

import {
  removePortraitOrientation,
  setLeanBackLauncherIntent,
  setTVBanner,
} from '../withTVAndroidManifest';
import { addTVPodfileModifications } from '../withTVPodfile';
import { addTVSplashScreenModifications } from '../withTVSplashScreen';
import {
  createBrandAssetsAsync,
  SourceImageJson,
  SourceBrandAssetsJson,
} from '../utils';

const { readAndroidManifestAsync } = AndroidConfig.Manifest;

jest.mock('fs');

const projectRoot = '/wat';

describe('withTV iOS/tvOS tests', () => {
  beforeEach(() => {
    vol.reset();
  });
  test('Add TV Podfile changes', async () => {
    const modifiedPodfile = addTVPodfileModifications(originalPodfile);
    expect(modifiedPodfile).toMatchSnapshot();
  });
  test('Add TV splash screen changes', async () => {
    const modifiedSplashScreen =
      addTVSplashScreenModifications(originalSplashScreen);
    expect(modifiedSplashScreen).toMatchSnapshot();
  });
  test('Create Apple TV brand assets', async () => {
    vol.fromJSON(
      {
        'assets/images/icon.png': 'icon.png',
        'assets/images/iconSmall.png': 'iconSmall.png',
        'assets/images/topShelf.png': 'topShelf.png',
        'assets/images/topShelf2x.png': 'topShelf2x.png',
      },
      projectRoot,
    );
    const iconSourceImages: SourceImageJson[] = [
      {
        path: join(projectRoot, 'assets/images/iconSmall.png'),
        scale: '1x',
      },
      {
        path: join(projectRoot, 'assets/images/icon.png'),
        scale: '2x',
      },
    ];
    const topShelfSourceImages: SourceImageJson[] = [
      {
        path: join(projectRoot, 'assets/images/topShelf.png'),
        scale: '1x',
      },
      {
        path: join(projectRoot, 'assets/images/topShelf2x.png'),
        scale: '2x',
      },
    ];
    const sourceBrandAssets: SourceBrandAssetsJson = {
      name: 'TVAppIcon',
      assets: [
        {
          role: 'top-shelf-image',
          size: '1920x720',
          imageSet: {
            name: 'Top Shelf Image',
            sourceImages: topShelfSourceImages,
          },
        },
        {
          role: 'primary-app-icon',
          size: '400x240',
          imageStack: {
            name: 'App Icon',
            sourceLayers: [
              {
                name: 'Front',
                sourceImages: iconSourceImages,
              },
              {
                name: 'Middle',
                sourceImages: iconSourceImages,
              },
              {
                name: 'Back',
                sourceImages: iconSourceImages,
              },
            ],
          },
        },
      ],
    };
    await createBrandAssetsAsync(projectRoot, sourceBrandAssets);
    const topLevelContentJson = await fs.readFile(
      resolve(projectRoot, 'TVAppIcon.brandassets', 'Contents.json'),
      { encoding: 'utf-8' },
    );
    expect(topLevelContentJson).toMatchSnapshot();
    const appIconStackContentJson = await fs.readFile(
      resolve(
        projectRoot,
        'TVAppIcon.brandassets',
        'App Icon.imagestack',
        'Contents.json',
      ),
      { encoding: 'utf-8' },
    );
    expect(appIconStackContentJson).toMatchSnapshot();
    const appIconFrontLayerJson = await fs.readFile(
      resolve(
        projectRoot,
        'TVAppIcon.brandassets',
        'App Icon.imagestack',
        'Front.imagestacklayer',
        'Contents.json',
      ),
      { encoding: 'utf-8' },
    );
    expect(appIconFrontLayerJson).toMatchSnapshot();
    const appIconContents = await fs.readFile(
      resolve(
        projectRoot,
        'TVAppIcon.brandassets',
        'App Icon.imagestack',
        'Front.imagestacklayer',
        'Content.imageset',
        'icon.png',
      ),
      { encoding: 'utf-8' },
    );
    expect(appIconContents).toEqual('icon.png');
  });
});

describe('with TV Android tests', () => {
  beforeEach(() => {
    vol.reset();
  });
  test('Adds leanback launcher intent category for TV builds', async () => {
    vol.fromJSON(
      {
        'androidManifest.xml': originalAndroidManifest,
      },
      projectRoot,
    );
    const originalManifest = await readAndroidManifestAsync(
      resolve(projectRoot, 'androidManifest.xml'),
    );
    const modifiedManifest = setLeanBackLauncherIntent({}, originalManifest, {
      isTV: true,
      showVerboseWarnings: false,
    });
    expect(JSON.stringify(modifiedManifest).indexOf('LEANBACK')).not.toEqual(
      -1,
    );
  });
  test('Adds TV banner to main application', async () => {
    vol.fromJSON(
      {
        'androidManifest.xml': originalAndroidManifest,
      },
      projectRoot,
    );
    const originalManifest = await readAndroidManifestAsync(
      resolve(projectRoot, 'androidManifest.xml'),
    );
    const modifiedManifest = setTVBanner(
      {},
      originalManifest,
      {
        isTV: true,
        showVerboseWarnings: false,
      },
      'bogus',
    );
    expect(
      JSON.stringify(modifiedManifest).indexOf('android:banner'),
    ).not.toEqual(-1);
  });
  test('Throws if manifest has no main intent', async () => {
    vol.fromJSON(
      {
        'androidManifest.xml': originalAndroidManifestNoMainIntent,
      },
      projectRoot,
    );
    const originalManifest = await readAndroidManifestAsync(
      resolve(projectRoot, 'androidManifest.xml'),
    );
    try {
      setLeanBackLauncherIntent({}, originalManifest, {
        isTV: true,
        showVerboseWarnings: false,
      });
      // Should not reach this line
      expect(true).toBe(false);
    } catch (e) {
      expect(e.message).toContain(
        'no main intent in main activity of Android manifest',
      );
    }
  });
  test('Removes orientation from activity metadata for TV builds', async () => {
    vol.fromJSON(
      {
        'androidManifest.xml': originalAndroidManifest,
      },
      projectRoot,
    );
    const originalManifest = await readAndroidManifestAsync(
      resolve(projectRoot, 'androidManifest.xml'),
    );
    const modifiedManifest = removePortraitOrientation({}, originalManifest, {
      isTV: false,
      showVerboseWarnings: false,
    });
    expect(
      JSON.stringify(modifiedManifest).indexOf('screenOrientation'),
    ).toEqual(-1);
  });
});
