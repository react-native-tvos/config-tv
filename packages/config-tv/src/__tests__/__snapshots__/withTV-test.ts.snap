// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withTV iOS/tvOS tests Add TV Podfile changes 1`] = `
"# @generated begin react-native-tvos-import - expo prebuild (DO NOT MODIFY) sync-45ad7228312d07c1751c6969560a210e8ec46f80
source 'https://github.com/react-native-tvos/react-native-tvos-podspecs.git'
source 'https://cdn.cocoapods.org/'

# @generated end react-native-tvos-import

require 'json'
podfile_properties = JSON.parse(File.read(File.join(__dir__, 'Podfile.properties.json'))) rescue {}

platform :tvos, '13.4'

target 'Test' do
end
"
`;

exports[`withTV iOS/tvOS tests Add TV splash screen changes 1`] = `
"
<?xml version="1.0" encoding="UTF-8"?>
<document type="com.apple.InterfaceBuilder.AppleTV.Storyboard" version="3.0" toolsVersion="16096" targetRuntime="AppleTV" propertyAccessControl="none" useAutolayout="YES" launchScreen="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="EXPO-VIEWCONTROLLER-1">
    <device id="appleTV" orientation="portrait" appearance="light"/>
    <dependencies>
        <deployment identifier="tvOS"/>
        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="16087"/>
        <capability name="Safe area layout guides" minToolsVersion="9.0"/>
        <capability name="documents saved in the Xcode 8 format" minToolsVersion="8.0"/>
    </dependencies>
    <scenes>
        <scene sceneID="EXPO-SCENE-1">
            <objects>
                <viewController storyboardIdentifier="SplashScreenViewController" id="EXPO-VIEWCONTROLLER-1" sceneMemberID="viewController">
                    <view key="view" userInteractionEnabled="NO" contentMode="scaleToFill" insetsLayoutMarginsFromSafeArea="NO" id="EXPO-ContainerView" userLabel="ContainerView">
                        <rect key="frame" x="0.0" y="0.0" width="414" height="736"/>
                        <autoresizingMask key="autoresizingMask" flexibleMaxX="YES" flexibleMaxY="YES"/>
                        <subviews>
                            <imageView userInteractionEnabled="NO" contentMode="scaleAspectFill" horizontalHuggingPriority="251" verticalHuggingPriority="251" insetsLayoutMarginsFromSafeArea="NO" image="SplashScreenBackground" translatesAutoresizingMaskIntoConstraints="NO" id="EXPO-SplashScreenBackground" userLabel="SplashScreenBackground">
                                <rect key="frame" x="0.0" y="0.0" width="414" height="736"/>
                            </imageView>
                            <imageView id="EXPO-SplashScreen" userLabel="SplashScreen" image="SplashScreen" contentMode="scaleAspectFit" horizontalHuggingPriority="251" verticalHuggingPriority="251" clipsSubviews="true" userInteractionEnabled="false" translatesAutoresizingMaskIntoConstraints="false">
                                <rect key="frame" x="0" y="0" width="414" height="736"/>
                            </imageView>
                        </subviews>
                        <viewLayoutGuide key="safeArea" id="Rmq-lb-GrQ"/>
                    </view>
                </viewController>
                <placeholder placeholderIdentifier="IBFirstResponder" id="EXPO-PLACEHOLDER-1" userLabel="First Responder" sceneMemberID="firstResponder"/>
            </objects>
            <point key="canvasLocation" x="140.625" y="129.4921875"/>
        </scene>
    </scenes>
    <resources>
        <image name="SplashScreenBackground" width="1" height="1"/>
        <image name="SplashScreen" width="414" height="736"/>
    </resources>
</document>
"
`;

exports[`withTV iOS/tvOS tests Create Apple TV brand assets 1`] = `
"{
  "assets": [
    {
      "filename": "Top Shelf Image.imageset",
      "role": "top-shelf-image",
      "size": "1920x720",
      "idiom": "tv"
    },
    {
      "filename": "App Icon.imagestack",
      "role": "primary-app-icon",
      "size": "400x240",
      "idiom": "tv"
    }
  ],
  "info": {
    "version": 1,
    "author": "expo"
  }
}"
`;

exports[`withTV iOS/tvOS tests Create Apple TV brand assets 2`] = `
"{
  "layers": [
    {
      "filename": "Front.imagestacklayer"
    },
    {
      "filename": "Middle.imagestacklayer"
    },
    {
      "filename": "Back.imagestacklayer"
    }
  ],
  "info": {
    "version": 1,
    "author": "expo"
  }
}"
`;

exports[`withTV iOS/tvOS tests Create Apple TV brand assets 3`] = `
"{
  "info": {
    "version": 1,
    "author": "expo"
  }
}"
`;
